Have KModal pass through their 'submit'  and 'cancel' events to simplify other modals

jonboiser opened the pull request:
### Summary

1. Changes KModal to have its parent emit 'submit' or 'cancel' events if these listeners have not been set on KModal. The point of this is to remove some boilerplate like `@submit="$emit('submit')` in the components that wrap the base KModal, and to reduce the amount variations on these event names (e.g. `@close`, `@success`, etc.) that appear in the code (in support of future work towards #5320). This makes the events of something like AssignmentCopyModal the same as the basic KModal.
1. For some modals, remove some coupling to Vuex (moving towards a full decoupling of modals from vuex so they are more like simple forms or input elements).
1. Wrap the text in KButton in a span, so it doesn't add extra whitespace when KButton is set to be basic-link
1. Fix a warning coming from CoreMenu

### Reviewer guidance
1. The simplest change is to delete lines like `@submit="$emit('submit')` from wrapped KModals, which should just work. The more complicated handlers like `@submit="handleSubmit` where handleSubmit may also emit or have complicated side-effects may take a little more checking.

### References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->

…

----

### Contributor Checklist


PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

jonboiser committed:
In KModal, pass ‘emit’ and ‘close’ events through parent

jonboiser committed:
Make SideNav toggle on keyup.esc

Since it will also interecpt keyup.esc on the privacy modal

jonboiser committed:
Update tests

jonboiser committed:
Update modals to use passed-through events

jonboiser committed:
Fix CoreMenu warning

jonboiser committed:
Wrap kbutton label to remove extra whitespace when basic-link

indirectlylit commented on kolibri/core/assets/src/views/KModal.vue:
```diff
@@ -228,16 +228,26 @@
+          if (this.$listeners.submit) {
+            this.$emit('submit');
+          } else {
+            this.$parent.$emit('submit');
```
I believe it should be the responsibility of the wrapper component to define this behavior because having a child modify the behavior of its parent goes against the typical component hierarchy model.

I haven't tried this before, but I think [you can use `v-on="$listeners"`](https://vuejs.org/v2/api/#vm-listeners) like this:

```html
<!-- AssignmentDeleteModal.vue -->
<template>
  <KModal
    :title="modalTitle"
    :submitText="coachStrings.$tr('deleteAction')"
    :cancelText="coachStrings.$tr('cancelAction')"
    v-on="$listeners"
  >
    <p>{{ modalDescription }}</p>
    <p v-if="modalConfirmation">
      {{ modalConfirmation }}
    </p>
  </KModal>
</template>
```

Admittedly it only cuts down one line of boilerplate rather than two, but I think the trade-off for explicitness and parent component control is worth it.

A similar strategy can be used for creating custom wrappers around native HTML elements like `<input>`. 

ref: https://github.com/vuejs/vue/issues/5983

jonboiser replied:
Using `v-on=$listeners` seems to work, but is kind of odd since instead of the Wrapper's child (KModal) defining the event passthrough behavior it's now the parent (of the wrapper) defining the passthrough behavior, namely by getting the event names right.

```
<ModalUser>
    <!-- "submit" will go through, but not "cancel", since the typo -->
    <!-- In either case, ModalUser component will not work -->
    <WrappedModal @submit="handleSubmit" @carncel="handleCancel" />
        <KModal v-on="$listeners />
```

Another thing I liked about the `$parent.$emit` hack is that you don't get duplicated `$emit` events in the devtools

jonboiser replied:
Okay, I can try using `v-on=$listeners` in this refactor, which gives us, in my opinion, three benefits

1. Introduces an interesting new syntax into the codebase that might useful for other things.
1. Still gets rid of some boilerplate
1. Still encourages the consistent use of the events "submit" and "cancel" for all Modal-type components.

jonboiser committed:
Revert "In KModal, pass ‘emit’ and ‘close’ events through parent"

This reverts commit ec04c7fa270c228a31dd55bce01c96cbd49d6ab4.

jonboiser committed:
Merge remote-tracking branch 'origin/release-v0.12.y' into consistent-modals

jonboiser committed:
Update tests

indirectlylit reviewed (approved):
looks good to me, thank you!

indirectlylit commented on kolibri/core/assets/src/views/UpdateNotification.vue:
```diff
@@ -84,7 +85,7 @@
         }
         this.removeNotification(this.id);
-        this.$emit('closeModal');
+        this.$emit('submit');
```
is it necessary to emit anything here? I think the `v-on` will do this for you?

indirectlylit commented on kolibri/plugins/coach/assets/src/views/plan/GroupsPage/CreateGroupModal.vue:
```diff
@@ -82,21 +82,18 @@
           this.submitting = true;
           this.createGroup({ groupName: this.name, classId: this.classId }).then(() => {
-            this.$emit('success');
+            this.$emit('submit');
```
won't this emit submit twice? Once for `v-on="$listeners"` and once here?

jonboiser replied:
CreateGroupModal only emits 'submit' once. I think if you explicitly set a `@submit=customListener`, it sort of overwrites the `{submit: parentListener}` hidden in `v-on="$listeners`

jonboiser replied:
Interestingly, though, if you remove this `$emit`, the event from CreateGroupModal does not appear in the devtools, but it seems that it still goes through the GroupsPage, which is not really what I want since i want to wait till the promise is resolved before emitting 'submit'.

I'm just going to remove the `v-on="listeners"` for the handful of modals that might emit the same event asynchronously.

In the future, I would like to to remove async behavior API calls from modals, and delegate them to the parents of modals. So modals just synchronously emit events and data, basically.

indirectlylit replied:
yeah that seems reasonable.

another alternative is for modals that handle their own internal async code, to explicitly emit an event that's _not_ `submit` or `cancel`. For example:

```javascript
.then(() => this.$emit('finished'))
```

indirectlylit reviewed (commented):
actually - question:

It looks like in a number of places we might be emitting events twice? E.g. once immediately due to `v-on...`, and again inside a custom event handler?

jonboiser committed:
Add warning if KModal is missing listeners

jonboiser committed:
Add @cancel handlers

jonboiser commented:
I've changed this so that every KModal simply re-emits "cancel" and/or "submit". Also added a warning on KModal in case someone shows a button, but doesnt actually register a listener for the button click.
